File: pkg/hub/tempo/conformance_test.go
package tempo

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"testing"
)

// tempoConformanceComponents mirrors the calendar-component object used to
// build an instant in a named zone in conformance/tempo.json.
type tempoConformanceComponents struct {
	Year     int    `json:"year"`
	Month    int    `json:"month"`
	Day      int    `json:"day"`
	Hour     int    `json:"hour"`
	Minute   int    `json:"minute"`
	Second   int    `json:"second"`
	TimeZone string `json:"timeZone"`
}

// tempoConformanceCase mirrors one entry in conformance/tempo.json. Results are
// rendered as strings ('iso' -> UTC ISO-8601, 'date' -> YYYY-MM-DD, 'int' ->
// decimal) so the comparison is language-neutral.
type tempoConformanceCase struct {
	Op       string                      `json:"op"`
	Base     *tempoConformanceComponents `json:"base"`
	Other    *tempoConformanceComponents `json:"other"`
	Arg      int                         `json:"arg"`
	Input    string                      `json:"input"`
	Pattern  string                      `json:"pattern"`
	TimeZone string                      `json:"timeZone"`
	Render   string                      `json:"render"`
	Expected string                      `json:"expected"`
	Note     string                      `json:"note"`
}

type tempoConformanceFile struct {
	Cases []tempoConformanceCase `json:"cases"`
}

// TestTempoConformance executes the shared Go<->TS tempo fixtures against the
// real Go API. It is the Go half of the cross-runtime drift guard (plan 008);
// the TS half lives in sdk/tempo/tests/src/conformance.test.ts and reads the
// same JSON.
func TestTempoConformance(t *testing.T) {
	cases := loadTempoConformance(t)

	if len(cases) == 0 {
		t.Fatal("no tempo conformance cases loaded")
	}

	for _, tc := range cases {
		tc := tc

		t.Run(tc.Op+"/"+tc.Note, func(t *testing.T) {
			got := runTempoOp(t, tc)

			if got != tc.Expected {
				t.Fatalf("op %s = %s, want %s (%s)", tc.Op, got, tc.Expected, tc.Note)
			}
		})
	}
}

func runTempoOp(t *testing.T, tc tempoConformanceCase) string {
	t.Helper()

	switch tc.Op {
	case "addDays":
		return renderTempo(t, buildTempo(t, tc.Base).AddDays(tc.Arg), tc.Render)
	case "addWeeks":
		return renderTempo(t, buildTempo(t, tc.Base).AddWeeks(tc.Arg), tc.Render)
	case "addHours":
		return renderTempo(t, buildTempo(t, tc.Base).AddHours(tc.Arg), tc.Render)
	case "addMonths":
		return renderTempo(t, buildTempo(t, tc.Base).AddMonths(tc.Arg), tc.Render)
	case "addMonthsNoOverflow":
		return renderTempo(t, buildTempo(t, tc.Base).AddMonthsNoOverflow(tc.Arg), tc.Render)
	case "diffInMonths":
		return strconv.Itoa(buildTempo(t, tc.Base).DiffInMonths(buildTempo(t, tc.Other)))
	case "diffInYears":
		return strconv.Itoa(buildTempo(t, tc.Base).DiffInYears(buildTempo(t, tc.Other)))
	case "parseFromPattern":
		factory, err := NewFactory(WithTimezone(tc.TimeZone))

		if err != nil {
			t.Fatalf("factory %q: %v", tc.TimeZone, err)
		}

		parsed, err := factory.FromFormat(tc.Input, tc.Pattern)

		if err != nil {
			t.Fatalf("parseFromPattern(%q, %q): %v", tc.Input, tc.Pattern, err)
		}

		return renderTempo(t, parsed, tc.Render)
	default:
		t.Fatalf("unknown tempo conformance op: %s", tc.Op)

		return ""
	}
}

func buildTempo(t *testing.T, components *tempoConformanceComponents) Time {
	t.Helper()

	if components == nil {
		t.Fatal("tempo conformance case missing components")
	}

	factory, err := NewFactory(WithTimezone(components.TimeZone))

	if err != nil {
		t.Fatalf("factory %q: %v", components.TimeZone, err)
	}

	value, err := factory.Create(Components{
		Year:   components.Year,
		Month:  components.Month,
		Day:    components.Day,
		Hour:   components.Hour,
		Minute: components.Minute,
		Second: components.Second,
	})

	if err != nil {
		t.Fatalf("create %+v: %v", components, err)
	}

	return value
}

func renderTempo(t *testing.T, value Time, render string) string {
	t.Helper()

	switch render {
	case "iso":
		return value.ISOString()
	case "date":
		return value.DateString()
	default:
		t.Fatalf("unknown tempo render mode: %s", render)

		return ""
	}
}

func loadTempoConformance(t *testing.T) []tempoConformanceCase {
	t.Helper()

	_, thisFile, _, ok := runtime.Caller(0)

	if !ok {
		t.Fatal("cannot resolve conformance test file path")
	}

	path := filepath.Join(filepath.Dir(thisFile), "..", "..", "..", "conformance", "tempo.json")

	data, err := os.ReadFile(path)

	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}

	var file tempoConformanceFile

	if err := json.Unmarshal(data, &file); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}

	return file.Cases
}

File: sdk/tempo/tests/src/conformance.test.ts
import { readFileSync } from 'node:fs';

import { Tempo } from '@hara/sdk-tempo';
import { describe, expect, it } from 'vite-plus/test';

// Shared Go<->TS golden vectors. This is the TS half of the cross-runtime drift
// guard (plan 008); the Go half lives in pkg/hub/tempo/conformance_test.go and
// reads the same JSON. Each case builds an instant from calendar components in a
// named zone, applies one operation, and renders the result as a string.

interface TempoConformanceComponents {
	year: number;
	month: number;
	day: number;
	hour?: number;
	minute?: number;
	second?: number;
	timeZone: string;
}

interface TempoConformanceCase {
	op: string;
	base?: TempoConformanceComponents;
	other?: TempoConformanceComponents;
	arg?: number;
	input?: string;
	pattern?: string;
	timeZone?: string;
	render?: string;
	expected: string;
	note: string;
}

interface TempoConformanceFile {
	cases: TempoConformanceCase[];
}

const fixturePath = new URL('../../../../conformance/tempo.json', import.meta.url);
const fixture = JSON.parse(readFileSync(fixturePath, 'utf8')) as TempoConformanceFile;

const buildTempo = (components: TempoConformanceComponents): Tempo =>
	Tempo.create({
		day: components.day,
		hour: components.hour ?? 0,
		minute: components.minute ?? 0,
		month: components.month,
		second: components.second ?? 0,
		timeZone: components.timeZone,
		year: components.year,
	});

const render = (value: Tempo, mode: string | undefined): string => {
	switch (mode) {
		case 'iso':
			return value.toISOString();

		case 'date':
			return value.toDateString();

		default:
			throw new Error(`unknown tempo render mode: ${mode}`);
	}
};

const runTempoOp = (testCase: TempoConformanceCase): string => {
	const { arg, base, op, other } = testCase;

	switch (op) {
		case 'addDays':
			return render(buildTempo(base as TempoConformanceComponents).addDays(arg as number), testCase.render);

		case 'addWeeks':
			return render(buildTempo(base as TempoConformanceComponents).addWeeks(arg as number), testCase.render);

		case 'addHours':
			return render(buildTempo(base as TempoConformanceComponents).addHours(arg as number), testCase.render);

		case 'addMonths':
			return render(buildTempo(base as TempoConformanceComponents).addMonths(arg as number), testCase.render);

		case 'addMonthsNoOverflow':
			return render(buildTempo(base as TempoConformanceComponents).addMonthsNoOverflow(arg as number), testCase.render);

		case 'diffInMonths':
			return String(buildTempo(base as TempoConformanceComponents).diffInMonths(buildTempo(other as TempoConformanceComponents)));

		case 'diffInYears':
			return String(buildTempo(base as TempoConformanceComponents).diffInYears(buildTempo(other as TempoConformanceComponents)));

		case 'parseFromPattern':
			return render(Tempo.fromFormat(testCase.input as string, testCase.pattern as string, { timeZone: testCase.timeZone }), testCase.render);

		default:
			throw new Error(`unknown tempo conformance op: ${op}`);
	}
};

describe('tempo cross-runtime conformance', () => {
	expect(fixture.cases.length).toBeGreaterThan(0);

	for (const testCase of fixture.cases) {
		it(`${testCase.op}: ${testCase.note}`, () => {
			expect(runTempoOp(testCase), testCase.note).toBe(testCase.expected);
		});
	}
});
